### PR TITLE
trying to fix build

### PR DESCRIPTION
**Pull request by jpbochi**
This is an attempt to fix https://github.com/ProtonMail/WebClient/issues/17

This PR is still WIP.

### CODE REVIEW

File: bin/npm.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+FROM node:6.10.3
+RUN yarn global add npm@5 --no-progress
```
Comment (dhoko) [inline bin/npm.Dockerfile]: We don't use yarn

File: package.json [hunk 2]
```diff
@@ -66,9 +69,11 @@
   },
   "license": "MIT",
   "scripts": {
-    "postinstall": "bower install && ./tasks/privateDependencies.sh",
```
Comment (dhoko) [inline package.json]: Why did you remove this hook ?

File: bin/npm.Dockerfile (hunk 1 above)
Comment (jpbochi) [reply bin/npm.Dockerfile]: sorry about that. I really didn't wan to introduce an extra tool into the mix. 
yarn is only being used here to install npm 5 in the node image. Yarn proved to be the a simple and consistent way to install npm@5. Since it comes with the image, there's no need to install anything extra.

The whole thing is sort of a chicken-egg problem. I tried the steps at https://github.com/npm/npm#super-easy-install, but using an old npm to install a new npm is not reliable. I also tried many variations of the suggested `curl -L https://www.npmjs.com/install.sh | sh`. In addition to offering less control over the npm version being installed, it wasn't very reliable.

Thanks for reviewing this, by the way. And since I got your attention, let me ask you something. Which version of node are you using in production? The readme only says "node.js >= v6". Using this docker approach, you can pin it and guarantee a much more consistent run in CI or for anyone trying to contribute. I chose 6.10.3 as just a guess.

As a final note, let me say that I was the one that originally created this make+docker solution about 2 years ago. I built the integration with circle. At some point, it was dropped for some reason. If are interested, I could help fix that integration again. It was done in a private ProtonMail repo where the WebClient v3 was being built. In the copy of that here, the history got lost. You can ask @jasonstockman, though. I even chatted with them about the whole thing over slack at the time.

Comment (dhoko) [reply bin/npm.Dockerfile]:
```sh
$ npm -v && node -v    
5.4.0                  
v8.2.1                 
```

As we use node for the build process, you can use `nvm` to install both node and npm. I didn't have any issues to move from npm 3 to 5 via `npm i -g npm@latest`, but maybe it doesn't work everytime, idk.

ok I will ask thx

File: package.json (hunk 2 above)
Comment (jpbochi) [reply package.json]: two reasons. One, bower was not happy running inside npm inside a container. With the `--allow-root` option, it works.

The second reason is harder to overcome. The privateDependencies script uses git to clone stuff from github. When running inside the node docker image, that won't work. My change here was to move it to the Makefile, and run it directly. I'd be happy to find some alternative that works both ways. For example, the script could exit early if the dependencies seem to already be there.

Speaking of bower, are there any plans to remove it? In their official repo, they recommend "yarn and webpack for new front-end projects". I'm sure npm@5 and webpack would do just fine.

**Comment by jpbochi** [comment]
@dhoko I must say this PR is still a work in progress. I apologise for not making that clear in the description.

My objectives were to both be able to run the client locally and to have a working CI build (with circle or other tool). As it is now, none have been achieved yet. For both, I'd need some help from the current contributors. If you allow me to bother you with more questions, I have some.

- How are you testing the code before releasing it to production?
- Would you be against the re-introduction of CircleCI or something equivalent?
- Are there guidelines to use the dev environment? If I'm reading the tests correctly, they create new users all the time.

File: package.json (hunk 2 above)
Comment (dhoko) [reply package.json]: > For example, the script could exit early if the dependencies seem to already be there.

The goal of this script is to always install a new version. We don't want to do nothing if there is already a directiry inside vendor, that's why there is a `$ rm -r`

I hope to kill bower in a few month, today it works and that's not something we need to improve yet.

File: bin/npm.Dockerfile (hunk 1 above)
Comment (jpbochi) [reply bin/npm.Dockerfile]: the official node 8 images come with npm 5, removing the need to install it at all. I'll use that here.

File: package.json (hunk 2 above)
Comment (jpbochi) [reply package.json]: cool. That makes sense. I suppose always cloning is more straight forward then cloning or updating-if-already-cloned.

File: package.json [hunk 3]
```diff
@@ -1,11 +1,14 @@
 {
   "repository": {
     "type": "git",
-    "url": "https://github.com/ProtonMail/Angular.git"
+    "url": "https://github.com/ProtonMail/WebClient.git"
```
Comment (jpbochi) [inline package.json]: I was trying to remember what was the name of the private repo. Now, I do. :)

BTW, I see that the version deployed to https://mail.protonmail.com is v3.11.2 from Sep 1st. The code here is at 3.11.0. Is the private repo still being used? It makes hard to make public contributions this way.

File: package-lock.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "protonmail-web",
-  "version": "3.10.2",
+  "version": "3.11.0",
```
Comment (jpbochi) [inline package-lock.json]: this is a sign that package-lock.json was not in sync with package.json.

File: package.json (hunk 3 above)
Comment (dhoko) [reply package.json]: :grin: voui. +1
We will try to update the repo.

**Comment by dhoko** [comment]
> My objectives were to both be able to run the client locally and to have a working CI build (with circle or other tool).

youhou

:+1: 

> How are you testing the code before releasing it to production?

no CI :sob: 

> Would you be against the re-introduction of CircleCI or something

I WANT A CI :) 

> Are there guidelines to use the dev environment?

You mean for e2e ? Hum I cannot give you an answer yet, they don't work very well today.  We lunch them with a test API.

**Comment by jpbochi** [comment]
> I WANT A CI :)

I'll see what I can do about that. :)

**Comment by jpbochi** [comment]
@dhoko I'm trying to integrate with shippable, since they have a pretty decent support for docker _and_ a free tier for public repos.

I managed to have all dependencies installed, but I'm having some problem building the code before running the tests. My last attempt fails with the following message:

```
PhantomJS 2.1.1 (Linux 0.0.0) ERROR
  SyntaxError: Use of reserved word 'import'
  at /root/src/github.com/jpbochi/WebClient/build/vendor/purify.js:1
```

You should be able to see all builds at https://app.shippable.com/github/jpbochi/WebClient/dashboard

Can you tell me what am I missing?

**Comment by dhoko** [comment]
Yep I changed the version of DOMPurify, the src isn't compatible anymore (-> _the lib doesn't work with ProtonMail anymore_)
I created a compatibility version https://github.com/ProtonMail/DOMPurify/commit/b06b4802f14ee5e69e0359557a9867d9e32cb6bf 

I hope we will sync WebClient soon,

**Comment by jpbochi** [comment]
in my last commit, I changed the version to include in the build to be `vendor/dompurify/dist/purify.js`. With that, only 32 of 1065 unit tests failed locally.

I'm having some trouble with shippable (apparently, it doesn't like that my fork was renamed).

Update: I think I managed to get the build to run again. It's at https://app.shippable.com/github/jpbochi/ProtonMailWebClient/dashboard

**Comment by jpbochi** [comment]
Build is back on. Result was the same as my local run. Here's one of the failures.

```
PhantomJS 2.1.1 (Linux 0.0.0) signatureBuilder factory Forward message Insert signature ~ no proton signature action: forward isAfter: false should two empty lines before the signature FAILED
	Expected '<div><br></div><div><br></div><div class="protonmail_signature_block">
	                <div class="protonmail_signature_block-user"><strong>POPOPO</strong></div>
	                <div class="protonmail_signature_block-proton protonmail_signature_block-empty"></div>
	            </div><div><br></div><p>polo</p>' to match /(<div><br><\/div>){2}<div class="protonmail_signature_block "/.
	specs/message/factories/signatureBuilder.spec.js:2409:43
```

I'm not sure why it's failing. The expected HTML looks correct. Maybe it's some escaping problem with `<\/div>`. Ideas?

**Comment by dhoko** [comment]
No don't worry tests are broken with WebClient, I forgot to update them.
With the new version everything works :grimacing:

**Comment by jpbochi** [comment]
ha! Ok. I managed to fix them nevertheless. Commit incoming soon.

**Comment by jpbochi** [comment]
there you go, a green build: https://app.shippable.com/github/jpbochi/ProtonMailWebClient/runs/2/summary/

It runs lint and karma tests. I barely started to try to have the e2e tests running, but I realised that protractor is not even declared as a dev dependency. I suppose they can be introduced in CI later.

**Comment by jpbochi** [comment]
enabling shippable here should be very straight forward. And by the way, their free tier even allows 150 builds/month on private repos. See https://www.shippable.com/pricing.html

**Comment by jpbochi** [comment]
@dhoko hey. Are you still interested in merging this?

**Comment by dhoko** [comment]
Yup we're working on it :)

**Comment by dhoko** [comment]
I have a dockerFile to build the project, we will publish it soon.

**Comment by paxcodes** [comment]
This is probably stale? We should just close this?
